File: app/src/main/java/com/trevorwiebe/trackacow/presentation/fragments/ReportsFragment.java
package com.trevorwiebe.trackacow.presentation.fragments;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.presentation.activities.LotReportActivity;
import com.trevorwiebe.trackacow.domain.adapters.ReportsRecyclerViewAdapter;
import com.trevorwiebe.trackacow.domain.dataLoaders.main.lot.QueryLots;
import com.trevorwiebe.trackacow.data.db.entities.LotEntity;
import com.trevorwiebe.trackacow.domain.utils.Constants;
import com.trevorwiebe.trackacow.domain.utils.ItemClickListener;

import java.util.ArrayList;

public class ReportsFragment extends Fragment implements QueryLots.OnLotsLoaded {

    private static final String TAG = "ReportsFragment";

    public ReportsFragment() {
    }

    private ReportsRecyclerViewAdapter mReportsRvAdapter;
    private TextView mNoReportsTv;

    private ArrayList<LotEntity> mLotList = new ArrayList<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.fragment_reports, container, false);

        mNoReportsTv = rootView.findViewById(R.id.empty_reports_tv);

        RecyclerView reportsRv = rootView.findViewById(R.id.reports_rv);
        reportsRv.setLayoutManager(new LinearLayoutManager(getContext()));
        mReportsRvAdapter = new ReportsRecyclerViewAdapter();
        reportsRv.setAdapter(mReportsRvAdapter);

        reportsRv.addOnItemTouchListener(new ItemClickListener(getContext(), reportsRv, new ItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                Intent reportsIntent = new Intent(getContext(), LotReportActivity.class);
                String lotId = mLotList.get(position).getLotId();
                reportsIntent.putExtra("lotId", lotId);
                reportsIntent.putExtra("reportType", Constants.LOT);
                startActivity(reportsIntent);
            }

            @Override
            public void onLongItemClick(View view, int position) {

            }
        }));

        return rootView;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        new QueryLots(ReportsFragment.this).execute(context);
    }

    @Override
    public void onLotsLoaded(ArrayList<LotEntity> lotEntities) {
        mLotList = lotEntities;
        if (mLotList.size() == 0) {
            mNoReportsTv.setVisibility(View.VISIBLE);
        } else {
            mNoReportsTv.setVisibility(View.INVISIBLE);
        }
        mReportsRvAdapter.swapLotData(mLotList);
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/fragment_report/ReportsFragment.kt
package com.trevorwiebe.trackacow.presentation.fragment_report

import com.trevorwiebe.trackacow.domain.adapters.ReportsRecyclerViewAdapter
import android.widget.TextView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import com.trevorwiebe.trackacow.R
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.LinearLayoutManager
import com.trevorwiebe.trackacow.domain.utils.ItemClickListener
import android.content.Intent
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.trevorwiebe.trackacow.domain.models.lot.LotModel
import com.trevorwiebe.trackacow.presentation.lot_reports.LotReportActivity
import com.trevorwiebe.trackacow.domain.utils.Constants
import com.trevorwiebe.trackacow.domain.utils.DataSource
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class ReportsFragment : Fragment() {

    private var mReportsRvAdapter: ReportsRecyclerViewAdapter? = null
    private var mLotList: List<LotModel> = emptyList()
    private lateinit var mProgressBar: LinearProgressIndicator
    private lateinit var mNoReportsTv: TextView

    private val reportsFragmentViewModel: ReportsFragmentViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {

        val rootView = inflater.inflate(R.layout.fragment_reports, container, false)
        mNoReportsTv = rootView.findViewById(R.id.empty_reports_tv)
        val reportsRv = rootView.findViewById<RecyclerView>(R.id.reports_rv)
        mProgressBar = rootView.findViewById(R.id.reports_progress_bar)
        reportsRv.layoutManager = LinearLayoutManager(context)
        mReportsRvAdapter = ReportsRecyclerViewAdapter()
        reportsRv.adapter = mReportsRvAdapter
        reportsRv.addOnItemTouchListener(
            ItemClickListener(
                context,
                reportsRv,
                object : ItemClickListener.OnItemClickListener {
                    override fun onItemClick(view: View, position: Int) {
                        val reportsIntent = Intent(context, LotReportActivity::class.java)
                        val lotId = mLotList[position].lotPrimaryKey
                        val lotCloudDatabaseId = mLotList[position].lotCloudDatabaseId
                        reportsIntent.putExtra("lotId", lotId)
                        reportsIntent.putExtra("reportType", Constants.LOT)
                        reportsIntent.putExtra("lotCloudDatabaseId", lotCloudDatabaseId)
                        startActivity(reportsIntent)
                    }

                    override fun onLongItemClick(view: View, position: Int) {}
                })
        )

        lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                reportsFragmentViewModel.uiState.collect {

                    if (it.isFetchingFromCloud && it.dataSource == DataSource.Local && it.lotList.isEmpty()) {
                        mProgressBar.visibility = View.VISIBLE
                    } else {
                        mProgressBar.visibility = View.GONE
                    }

                    mLotList = it.lotList
                    if (mLotList.isEmpty()) {
                        Handler(Looper.getMainLooper()).postDelayed(
                            {
                                if (mLotList.isEmpty())
                                    mNoReportsTv.visibility = View.VISIBLE
                            }, 100 // value in milliseconds
                        )
                    } else {
                        mNoReportsTv.visibility = View.INVISIBLE
                    }
                    mReportsRvAdapter!!.swapLotData(mLotList)
                }
            }
        }

        return rootView
    }
}